File: src/main/scala/examples/GreetingActor_2.scala
package examples

import akka.actor._

case class SayHello(name: String)

class GreetingActor_2 extends Actor {

  def receive = {
    case hello: SayHello => {
      println("Hello " + hello.name)
      sender ! hello.name
    }
  }
}

object GreetingActor_2 extends App {
  val system = ActorSystem("MySystem")

  val greetingActor = system.actorOf(Props[GreetingActor_2])
  greetingActor ! SayHello("Pope Benedict")

  // There are better ways to ensure message are received before termination
  Thread.sleep(100)
  system.shutdown()
}

File: src/main/java/examples/GreetingActor_2.java
package examples;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.japi.pf.ReceiveBuilder;
import scala.PartialFunction;
import scala.runtime.BoxedUnit;

public class GreetingActor_2 extends AbstractActor {

    static class SayHello {
        String name;

        public SayHello(String n){
            name = n;
        }
    }

    @Override
    public PartialFunction<Object, BoxedUnit> receive() {
        return ReceiveBuilder.match(SayHello.class, hello -> {
            System.out.println("Hello " + hello.name);
            sender().tell(hello.name, self());
        }).build();
    }

    public static void main(String[] args) throws InterruptedException {
        ActorSystem system = ActorSystem.create("MySystem");

        ActorRef greetingActor = system.actorOf(Props.create(GreetingActor_2.class), "greeter");
        greetingActor.tell(new SayHello("Pope Benedict"), ActorRef.noSender());

        // There are better ways to ensure message are received before termination
        Thread.sleep(100);
        system.shutdown();
    }
}
